certora/specs/ValidatorsExitBusOracle.spec

```diff
@@ -0,0 +1,355 @@
+
+definition excludeHelper(method f) returns bool =
+    f.selector != submitReportDataHelper(uint256, uint256, uint256, uint256, bytes, uint256).selector;
+
+
+definition matchReportDataWithGlobal(uint256 consensusVersion, uint256 refSlot, uint256 requestsCount, uint256 dataFormat) returns bool = 
+    consensusVersion == getConsensusVersionGlobal()
+    && refSlot == getRefSlotGlobal()
+    && requestsCount == getRequestsCountGlobal()
+    && dataFormat == getDataFormatGlobal()
+    // && compareBytes(dataInput, dataGlobal)
+    ;
+
+
+// STATUS - in progress (can be violated becuase admin can remove it's own role)
+// admin != 0
+invariant adminIsNotZero(address admin)
+    hasRole(DEFAULT_ADMIN_ROLE(), admin) => admin != 0
+
+
+// STATUS - verified
+//  It must not be possible to initialize twice.
+rule cantInitializeTwice(env e, env e2, method f, method g) 
+    filtered { f -> f.selector == initialize(address, address, uint256, uint256).selector,
+               g -> g.selector == initialize(address, address, uint256, uint256).selector } {
+
+    calldataarg args;
+    calldataarg args2;
+
+    f(e, args);
+    g@withrevert(e2, args2);
+
+    assert lastReverted, "Remember, with great power comes great responsibility.";
+}
+
+
+// STATUS - verified
+// Only a user with necessary role can resume within pause timeframe
+rule canResume(env e, env e2) {
+    bool pausedBefore = isPaused(e);
+
+    require e.msg.sender == e2.msg.sender;
+
+    resume@withrevert(e2);
+    bool isReverted = lastReverted;
+
+    bool pausedAfter = isPaused(e2);
+
+    assert (e2.msg.value == 0 
+                && pausedBefore 
+                && e2.block.timestamp < getResumeSinceTimestamp()
+                && hasRole(RESUME_ROLE(), e2.msg.sender))
+            => (!isReverted 
+                && !pausedAfter);
+}
+
+
+// STATUS - verified
+// It must not be possible to submit the same report twice. 
+rule cantSubmitReportTwice(env e, env e2, method f, method g) {
+    uint256 contractVersion1;    uint256 contractVersion2;
+    uint256 consensusVersion1;   uint256 consensusVersion2;
+    uint256 refSlot1;            uint256 refSlot2;
+    uint256 requestsCount1;      uint256 requestsCount2;
+    uint256 dataFormat1;         uint256 dataFormat2;
+    bytes   dataInput1;          bytes   dataInput2;
+
+    require matchReportDatas(consensusVersion1, refSlot1, requestsCount1, dataFormat1, dataInput1);
+    submitReportDataHelper(e, consensusVersion1, refSlot1, requestsCount1, dataFormat1, dataInput1, contractVersion1);
+    
+    require matchReportDatas(consensusVersion2, refSlot2, requestsCount2, dataFormat2, dataInput2);
+    submitReportDataHelper@withrevert(e2, consensusVersion2, refSlot2, requestsCount2, dataFormat2, dataInput2, contractVersion2);
+    bool isReverted = lastReverted;
+
+    assert (e.block.timestamp == e2.block.timestamp 
+                || (contractVersion1 == contractVersion2 
```

Comment:
When you call require the 'matchReportDatas for both reports`, you already require all those values to be equal to each other, right?
So all those equalities in the assert are redundant.
Did I miss something?

Reply:
It's related to harness not to both reports I used here.
I had to create a global report value because the one that passed as a function argument randomly altered by the tool. I asked your help with harnessing of assembly block if you remember